### PR TITLE
GA-181 p4 with retry approach

### PR DESCRIPTION
**Pull request by abul-syed**
### Jira link (if applicable)

https://tools.hmcts.net/jira/browse/GA-181

### Change description ###

Subtask to update process 4 with retry mechanism mentioned in parent ticket [GA-172](https://tools.hmcts.net/jira/browse/GA-172)

### CODE REVIEW

File: src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/ProfessionalUserIntegrationTest.java [hunk 1]
```diff
@@ -78,4 +85,58 @@ void shouldInsertOneUserIntoUserRefreshQueue_AndClearOrganisationRefreshQueue()
...
+        when(prdService.fetchUsersByOrganisation(any(), eq(null), eq(null), any()))
+                .thenReturn(ResponseEntity.ok(page1));
+
+        // throwing exception on 2nd page retrieval to test rollback (user inserts from page 1 SHOULD be rolled back)
+        when(prdService.fetchUsersByOrganisation(any(), any(String.class), any(String.class), any()))
+                .thenThrow(ServiceException.class);
+
+        professionalUserService.findAndInsertUsersWithStaleOrganisationsIntoRefreshQueue();
+
+        assertEquals(userRefreshQueueRepository.findAll().size(), 0);
```
Comment (danlysiak) [inline src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/ProfessionalUserIntegrationTest.java:107]:
```suggestion
        assertEquals(0, userRefreshQueueRepository.findAll().size());
```

File: src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/ProfessionalUserIntegrationTest.java [hunk 2]
```diff
@@ -78,4 +85,58 @@ void shouldInsertOneUserIntoUserRefreshQueue_AndClearOrganisationRefreshQueue()
...
```
Comment (danlysiak) [inline src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/ProfessionalUserIntegrationTest.java:134]:
```suggestion
        assertEquals(0, userRefreshQueueRepository.findAll().size());
```

File: src/main/resources/application.yaml [hunk 3]
```diff
@@ -197,4 +197,7 @@ professional:
       scheduling:
         enabled: ${PROFESSIONAL_SCHEDULING_ENABLED:false}
         findUsersWithStaleOrganisations:
-          cron: ${CRON_P4_USERS_STALE_ORGS:*/30 * * * * *}
\ No newline at end of file
+          cron: ${CRON_P4_USERS_STALE_ORGS:*/30 * * * * *}
+      retryOneInterval: ${RETRY_ONE_INTERVAL:2}
+      retryTwoInterval: ${RETRY_TWO_INTERVAL:15}
+      retryThreeInterval: ${RETRY_THREE_INTERVAL:60}
```
Comment (danlysiak) [inline src/main/resources/application.yaml:203]: Could consider putting these inside the `findUsersWithStaleOrganisations` so we can independently configure the retries across this and the other process(es) that need retries it in future. Perhaps also a comment in here just to say that it's in minutes so anyone looking back at this in future doesn't need to check the implementation.

**Review by danlysiak — APPROVED** [review]
Minor optional comment but otherwise looking good 😄

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/OrganisationRefreshQueueRepository.java [hunk 4]
```diff
@@ -28,4 +31,24 @@ public interface OrganisationRefreshQueueRepository extends JpaRepository<Organi
...
+    @Query(value = "update organisation_refresh_queue "
+            + "set "
+                + "retry = case "
+                    + "when retry = 0 then 1 "
+                    + "when retry = 1 then 2 "
+                    + "when retry = 2 then 3 "
+                    + "else 4 "
+                + "end, "
+                + "retry_after = case "
+                    + "when retry = 0 then now() + (interval '1' Minute) * :retryOneInterval "
```
Comment (sanjayparekh-hmcts) [inline src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/OrganisationRefreshQueueRepository.java:46]: maybe its just easier to pass the interval parms into the interval clause - I did that on one of my changes

File: src/main/resources/application.yaml (hunk 3 above)
Comment (sanjayparekh-hmcts) [reply src/main/resources/application.yaml:203]: maybe its worth labelling the key as retryOneIntervalMinutes...

**Review by sanjayparekh-hmcts — CHANGES_REQUESTED** [review]

File: src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/ProfessionalUserIntegrationTest.java (hunk 1 above)
Comment (abul-syed) [reply src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/ProfessionalUserIntegrationTest.java]: suggestion applied in [ce66cc0](https://github.com/hmcts/am-org-role-mapping-service/pull/1788/commits/ce66cc0e2224ee0c89290c061f894bf27e86e0dc)

File: src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/ProfessionalUserIntegrationTest.java (hunk 2 above)
Comment (abul-syed) [reply src/integrationTest/java/uk/gov/hmcts/reform/orgrolemapping/domain/service/ProfessionalUserIntegrationTest.java]: suggestion applied in [ce66cc0](https://github.com/hmcts/am-org-role-mapping-service/pull/1788/commits/ce66cc0e2224ee0c89290c061f894bf27e86e0dc)

File: src/main/resources/application.yaml (hunk 3 above)
Comment (abul-syed) [reply src/main/resources/application.yaml]: moved under `findUsersWithStaleOrganisations` & updated name to `retryOneIntervalMin` in [a5edcef](https://github.com/hmcts/am-org-role-mapping-service/pull/1788/commits/a5edcefe4b742fde986d16b64c99503a97c643f2)

File: src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/OrganisationRefreshQueueRepository.java (hunk 4 above)
Comment (abul-syed) [reply src/main/java/uk/gov/hmcts/reform/orgrolemapping/data/OrganisationRefreshQueueRepository.java]: i tried that initially but kept getting error:

`Caused by: org.postgresql.util.PSQLException: ERROR: invalid input syntax for type interval: ":retryOneInterval Minute"`

i think its not possible, see SO: https://stackoverflow.com/questions/43767284/postgres-interval-not-working-with-native-spring-data-jpa-query/43767356#43767356

**Pull request merged**